File: day2/day2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func preprocesLine(line string) (int, int, byte, string) {
	splitLine := strings.Split(line, " ")
	password := splitLine[2]
	letter := splitLine[1][0]

	splitMinMax := strings.Split(splitLine[0], "-")
	min, _ := strconv.Atoi(splitMinMax[0])
	max, _ := strconv.Atoi(splitMinMax[1])

	return min, max, letter, password
}

func isPasswordValidPart1(min int, max int, letter byte, password string) bool {
	count := strings.Count(password, string(letter))
	if count < min || count > max {
		return false
	}
	return true
}

func isPasswordValidPart2(min int, max int, letter byte, password string) bool {
	valid := false
	if password[min-1] == letter {
		valid = true
	}
	if password[max-1] == letter {
		valid = !valid
	}
	return valid
}

func main() {
	file, _ := os.Open("day2_input.txt")

	validPasswordsP1 := 0
	validPasswordsP2 := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		min, max, letter, password := preprocesLine(line)
		if isPasswordValidPart1(min, max, letter, password) {
			validPasswordsP1++
		}
		if isPasswordValidPart2(min, max, letter, password) {
			validPasswordsP2++
		}
	}
	fmt.Println(validPasswordsP1)
	fmt.Println(validPasswordsP2)
}

File: day2/day2.py
example_input = ["1-3 a: abcde", "1-3 b: cdefg", "2-9 c: ccccccccc"]

def preprocess_line(line):
    split_line = line.split(" ")
    low, high = map(int, split_line[0].split("-"))
    letter = split_line[1][0]
    password = split_line[2]
    return low, high, letter, password

def is_valid_password_part2(low, high, letter, password):
    valid = False
    if password[low-1] == letter: 
        valid = True
    if password[high-1] == letter: 
        valid = not valid
    return valid

def main():
    valid_passwords_p1 = 0
    valid_passwords_p2 = 0
    with open("day2.in") as file:
        for line in file:
            low, high, letter, password = preprocess_line(line)
            valid_passwords_p1 += low <= password.count(letter) <= high
            if is_valid_password_part2(low, high, letter, password):
                valid_passwords_p2 += 1
    print(valid_passwords_p1)
    print(valid_passwords_p2)
main()